datafusion/expr-common/src/groups_accumulator.rs

```diff
@@ -52,8 +71,248 @@ impl EmitTo {
+        mode: GroupStatesMode,
+    ) -> Vec<T> {
+        match self {
+            Self::All => {
+                debug_assert!(matches!(mode, GroupStatesMode::Flat));
+                blocks.pop_first_block().unwrap()
+            }
+            Self::First(n) => {
+                debug_assert!(matches!(mode, GroupStatesMode::Flat));
+
+                let block = blocks.current_mut().unwrap();
+                let split_at = min(block.len(), *n);
+
+                // get end n+1,.. values into t
+                let mut t = block.split_off(split_at);
+                // leave n+1,.. in v
+                std::mem::swap(block, &mut t);
+                t
+            }
+            Self::NextBlock(_) => {
+                debug_assert!(matches!(mode, GroupStatesMode::Blocked(_)));
+                blocks.pop_first_block().unwrap()
+            }
+        }
+    }
+}
+
+/// Mode for `accumulators` and `group values`
+///
+/// Their meanings:
+///   - Flat, the values in them will be managed with a single `Vec`.
+///     It will grow constantly when more and more values are inserted,
+///     that leads to a considerable amount of copying, and finally a bad performance.
+///
+///   - Blocked(block_size), the values in them will be managed with multiple `Vec`s.
+///     When the block is large enough(reach block_size), a new block will be allocated
+///     and used for inserting.
+///     Obviously, this strategy can avoid copying and get a good performance.
+#[derive(Debug, Clone, Copy)]
+pub enum GroupStatesMode {
+    Flat,
+    Blocked(usize),
+}
+
+/// Blocked style group index used in blocked mode group values and accumulators
+///
+/// Parts in index:
+///   - High 32 bits represent `block_id`
+///   - Low 32 bits represent `block_offset`
+#[derive(Debug, Clone, Copy)]
+pub struct BlockedGroupIndex {
+    pub block_id: usize,
+    pub block_offset: usize,
+}
+
+impl BlockedGroupIndex {
+    pub fn new(group_index: usize) -> Self {
+        let block_id =
+            ((group_index as u64 >> 32) & BLOCKED_INDEX_LOW_32_BITS_MASK) as usize;
+        let block_offset =
+            ((group_index as u64) & BLOCKED_INDEX_LOW_32_BITS_MASK) as usize;
+
+        Self {
+            block_id,
+            block_offset,
         }
     }
+
+    pub fn new_from_parts(block_id: usize, block_offset: usize) -> Self {
+        Self {
+            block_id,
+            block_offset,
+        }
+    }
+
+    pub fn as_packed_index(&self) -> usize {
+        ((((self.block_id as u64) << 32) & BLOCKED_INDEX_HIGH_32_BITS_MASK)
+            | (self.block_offset as u64 & BLOCKED_INDEX_LOW_32_BITS_MASK))
+            as usize
+    }
+}
+
+/// The basic data structure for blocked aggregation intermediate results
+pub struct Blocks<T> {
+    /// The current block, it should be pushed into `previous`
+    /// when next block is pushed
+    current: Option<T>,
+
+    /// Previous blocks pushed before `current`
+    previous: VecDeque<T>,
```

Comment:
What is the reason there is `current` and `previous` instead of one `VecDeque<T>`? Is it for performance concern?

Reply:
> What is the reason there is `current` and `previous` instead of one `VecDeque<T>`?

It is a small optimzation especially for single case.

For example, when blocked optimization is disabled(e.g. in streaming aggregation).
- The `Block<Vec<T>>` in accumulator will just the the `current`, and `previous` will be always empty.
- We will only call the `current()` and `current_mut` of `Block<Vec<T>>`, and it will just simply return the reference of the inner current(a `Vec<T>`).
- We can avoid use the `back` and `back_mut` which will is a bit more expansive than return `Vec<T>` directly.

Reply:
But I suspect it may be actually caused by https://github.com/apache/datafusion/pull/11943#discussion_r1723425840.